src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -76,34 +79,56 @@ public void run() {
             phaser.arrive();
             while (true) {
                 try {
-                    PublishedMetaDbRequest request = cmoLabelGeneratorQueue.poll(100, TimeUnit.MILLISECONDS);
-                    if (request != null) {
-                        // skip request if filtering by only cmo requests and cmoRequest status is false
-                        if (igoCmoRequestFilter && !request.isCmoRequest()) {
-                            requestStatusLogger.logRequestStatus(request.getRequestJson(),
-                                    RequestStatusLogger.StatusType.CMO_REQUEST_FILTER_SKIPPED_REQUEST);
-                            continue;
-                        }
+                    String requestJson = cmoLabelGeneratorQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (requestJson != null && requestJson.length() > 10) {
                         // skip request if there are no samples to generate cmo labels for
-                        if (request.getSamples().isEmpty()) {
-                            requestStatusLogger.logRequestStatus(request.getRequestJson(),
+                        List<SampleMetadata> samples = getSamplesFromRequestJson(requestJson);
+                        if (samples.isEmpty()) {
+                            requestStatusLogger.logRequestStatus(requestJson,
                                     RequestStatusLogger.StatusType.REQUEST_WITH_MISSING_SAMPLES);
                             continue;
                         }
+                        Map<String, List<SampleMetadata>> patientSamplesMap = new HashMap<>();
+
                         // generate cmo labels for each sample metadata
                         List<SampleMetadata> updatedSamples = new ArrayList<>();
-                        for (SampleMetadata sample : request.getSamples()) {
-                            // generate cmo label
-                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample);
+                        for (SampleMetadata sample : samples) {
+                            // get or request existing patient samples and update patient sample mapping
+                            if (!patientSamplesMap.containsKey(sample.getCmoPatientId())) {
+                                Message reply = messagingGateway.request(PATIENT_SAMPLES_REQUEST_TOPIC,
```

Comment:
In an effort to cut down the size of this loop, maybe introduce a method that like SampleMetadata[] getPatientSamples(String cmoPatientId);

Also, I mentioned to @divyamadala30  that we have to change the signature of request().  We should not be returning a Message instance, which is NATs specific, we should return Object or byte[].  We should update the code in anticipation of this change to Gateway interface.

Reply:
Ah, we talked about that a bit yesterday. We're already handling Message objects in the message consumers due to our having to manually acknowledge messages so I was a little hesitant to change the returned type here. We can discuss more during standup.